### PR TITLE
kubeswitch: fix broken package; add nixos module

### LINKED ISSUE #237886: Kubeswitch pkg missing the script that makes it work
### Describe the bug
Kubeswitch requires both the binary and a script https://github.com/danielfoehrKn/kubeswitch/blob/master/docs/installation.md#switch-command

### Steps To Reproduce
1. Install kubeswitch
2. Try to use `switch`

### Expected behavior
Kubeconfig is set as intended

### Additional context
https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/development/tools/kubeswitch/default.nix only adds the binary, not the script itself.

On the other hand, I'm not using nix directly but via fleek.

### Notify maintainers
@bryanasdev000

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 22.5.0, macOS 13.4`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/store/fqjg9y9pwh9cg2izs76kkv0wwwpsqp1v-source`
```

### PR DESCRIPTION
**Pull request by 1nv0k32**
## Description of changes
Fix the package and add the program to set up the shell to use the kubeswitch properly.
P.S: The command `switch` was a terrible choice by the developer as it has lots of collision with other commands, so I made a program and added a field `commandName` to make it customizable. By default, this field is set to `kswitch`.

closes #237886

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by 1nv0k32** [comment]
@bryanasdev000 
Sorry for closing the other PR as it was not complete and I've found issues in shell completions and also in the init script.
This [Link](https://github.com/danielfoehrKn/kubeswitch/blob/master/docs/how_it_works.md) explains how it should be set up and work.

**Comment by 1nv0k32** [comment]
@bryanasdev000 
Do you time to review this PR?
Thanks in advance.

**Comment by bryanasdev000** [comment]
> @bryanasdev000 Do you time to review this PR? Thanks in advance.

Sorry for the delay, busy with RL.

I have tested the package, and it worked. When I arrive home, I will patch my local nixpkgs with your PR and test the module too.

Thanks.

File: nixos/modules/programs/kubeswitch.nix
Comment (bryanasdev000) [inline nixos/modules/programs/kubeswitch.nix:49]: Same here.

**Review by bryanasdev000 — APPROVED** [review]
Sorry for the late response. I tested today with a few kind clusters and worked flawless, just added a few nitpicks for shell.

I'm not sure how PRs with modules are going, but you may need to add a note to `nixos/doc/manual/release-notes/rl-2405.section.md`, also if willing, you can contribute to home-manager(https://github.com/nix-community/home-manager/) and have this module working in other distros too.

Many thanks.

EDIT: Also tested shell completion with ZSH, all working.

File: pkgs/development/tools/kubeswitch/default.nix
Comment (superherointj) [inline pkgs/development/tools/kubeswitch/default.nix:39]:
```suggestion
  meta = {
```

**Comment by superherointj** [comment]
Please, split the commits between package and module following the usual conventions.

File: nixos/modules/programs/kubeswitch.nix
Comment (superherointj) [inline nixos/modules/programs/kubeswitch.nix:3]:
```suggestion
```

File: nixos/modules/programs/kubeswitch.nix
Comment (superherointj) [inline nixos/modules/programs/kubeswitch.nix:13]:
```suggestion
        commandName = mkOption {
```

**Comment by superherointj** [comment]
ping @1nv0k32

File: nixos/modules/programs/kubeswitch.nix
Comment (1nv0k32) [reply nixos/modules/programs/kubeswitch.nix:45]: I didn't do that because I think people might have other shells installed but not as the main shell.

File: nixos/modules/programs/kubeswitch.nix
Comment (1nv0k32) [reply nixos/modules/programs/kubeswitch.nix:3]: I've seen it done in other packages as it makes fewer repetitions of `lib` in the whole file.

File: nixos/modules/programs/kubeswitch.nix
Comment (1nv0k32) [reply nixos/modules/programs/kubeswitch.nix]: Done :D

**Comment by 1nv0k32** [comment]
@superherointj Sorry for the long delay as life was bit busy poking me :D

**Comment by 1nv0k32** [comment]
Ping @superherointj @bryanasdev000 😊

File: nixos/modules/programs/kubeswitch.nix
Comment (itzwam) [reply nixos/modules/programs/kubeswitch.nix:3]: or you can inherit the functions you're using specifically

Comment (1nv0k32) [reply nixos/modules/programs/kubeswitch.nix:3]: @itzwam Can you elaborate more please? Or a suggestion for a change?
Thanks in advance :D

Comment (superherointj) [reply nixos/modules/programs/kubeswitch.nix:3]: `nested with` can create issues in static analysis and can cause issues that are hard to diagnose. I'd recommend removing it.

**Comment by superherointj** [comment]
* Commit messages needs improvement:
  - NixOS module commit should be: `nixos/kubeswitch: init`
  - `kubeswitch` package commit should be: `kubeswitch: fix package` (You can expand the `subtitle` or add more information to the body of commit message. I'm trying to convey the format. Not the content.)
  - Both commit messages should be in `lowercase`.

* Consider using [nixfmt-rfc-style](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=nixfmt) for formatting of files. You might find it useful.

File: pkgs/development/tools/kubeswitch/default.nix
Comment (superherointj) [inline pkgs/development/tools/kubeswitch/default.nix:29]:
```suggestion
      $out/bin/switcher --cmd switcher completion $shell > switcher.$shell
```
Empty space

**Comment by superherointj** [comment]
There are differences between `nixfmt`(legacy) and `nixfmt-rfc-style` (current). I think you have used `nixfmt`. Please use `nixfmt-rfc-style` for formatting instead.

**Review by superherointj — APPROVED** [review]

**Comment by 1nv0k32** [comment]
Thanks for all the help, @superherointj 

May I ask what the next step is? After CI, who is gonna merge? Is it gonna end up in unstable, or can it make it to the next stable release?

**Comment by superherointj** [comment]
> Thanks for all the help, @superherointj

Welcome to nixpkgs!

> May I ask what the next step is? After CI, who is gonna merge?

The wait is for CI. Then, I can manually merge it.

> Is it gonna end up in unstable

It will reach `master` first (no cache). Then, `nixos-unstable-small` (partial cache). Then, `nixos-unstable` (cached).

You can track here: https://nixpk.gs/pr-tracker.html?pr=288162

> or can it make it to the next stable release?

Next release, yes.

**Comment by hedefalk** [comment]
kswitch was also a bad name :)

https://web.mit.edu/kerberos/www/krb5-latest/doc/user/user_commands/kswitch.html

```
🐟 kswitch --version
kswitch (Heimdal 1.5.1apple1)
Copyright 1995-2011 Kungliga Tekniska Högskolan
Send bug-reports to heimdal-bugs@h5l.org

🐟 which kswitch
/usr/bin/kswitch
```

**Comment by superherointj** [comment]
> kswitch was also a bad name :)
> 
> https://web.mit.edu/kerberos/www/krb5-latest/doc/user/user_commands/kswitch.html
> 
> ```
> 🐟 kswitch --version
> kswitch (Heimdal 1.5.1apple1)
> Copyright 1995-2011 Kungliga Tekniska Högskolan
> Send bug-reports to heimdal-bugs@h5l.org
> 
> 🐟 which kswitch
> /usr/bin/kswitch
> ```

Please, don't resurrect past PRs. Move forward. Propose the solution you consider right.

I am not a user of such solution. Only reviewed because PRs don't get reviewed. In case, you have an interest in such package consider becoming a maintainer. And then, leading it to the direction you consider right.

**Comment by superherointj** [comment]
I guess it should have been "kubeswitch" then.

**Comment by 1nv0k32** [comment]
@hedefalk Well you can choose the command name with the option `commandName` in the module.

**Comment by hedefalk** [comment]
> @hedefalk Well you can choose the command name with the option `commandName` in the module.

Yeah, I saw it, thanks!

**Pull request merged**
